### PR TITLE
Upgrade to lucene-6.4.0-snapshot-ec38570

### PR DESCRIPTION
**Pull request by jimczi**
The upgrade is needed to enable index sorting on multi-valued fields (all fields are multi-valued in ES).
Note for reviewers: I tried to split the commits in order to fix one problem at a time so better to check the diffs commit by commit (there are only 4 and the last 2 are real fixes).

`gradle check` passes on my machine (with x-plugins), let's see what the CI says and I'll also rerun the checks locally.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java [hunk 1]
```diff
@@ -619,6 +623,10 @@ public boolean escape() {
      * Default is <code>{@value #DEFAULT_SPLIT_ON_WHITESPACE}</code>.
      */
     public QueryStringQueryBuilder splitOnWhitespace(boolean value) {
+        if (autoGeneratePhraseQueries() && value == false) {
+            throw new IllegalArgumentException("it is disallowed to disable split_on_whitespace " +
+                "if auto_generate_phrase_queries is activated");
+        }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: Should we fail in `doToQuery` instead? I am wondering that the builder might be in an invalid state temporarily but in an ok state in the end depending on the order in which setters are called?

**Review by jpountz — APPROVED** [review]
I left one question, otherwise it looks good!

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java (hunk 1 above)
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: Yes good idea. I'll update this part.

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java [hunk 2]
```diff
@@ -937,6 +937,10 @@ protected int doHashCode() {
     protected Query doToQuery(QueryShardContext context) throws IOException {
         //TODO would be nice to have all the settings in one place: some change though at query execution time
         //e.g. field names get expanded to concrete names, defaults get resolved sometimes to settings values etc.
+        if (splitOnWhitespace == false && autoGeneratePhraseQueries) {
+            throw new IllegalArgumentException("it is disallowed to disable split_on_whitespace " +
+                "if auto_generate_phrase_queries is activated");
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: can you put brackets around configuration options to be consistent with most of our error messages?

**Review by jpountz — APPROVED** [review]
LGTM

**Pull request merged**